x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownReadinessIT.java

```diff
@@ -0,0 +1,89 @@
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.readiness.MockReadinessService;
+import org.elasticsearch.readiness.ReadinessService;
+import org.elasticsearch.test.ESIntegTestCase;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import static org.elasticsearch.readiness.MockReadinessService.tcpReadinessProbeTrue;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.empty;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class NodeShutdownReadinessIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getMockPlugins() {
+        final List<Class<? extends Plugin>> plugins = new ArrayList<>(super.getMockPlugins());
+        plugins.add(MockReadinessService.TestPlugin.class);
+        return Collections.unmodifiableList(plugins);
+    }
+
+    private void putNodeShutdown(String nodeId, SingleNodeShutdownMetadata.Type type, TimeValue allocationDelay) {
+        assertAcked(
+            client().execute(
+                PutShutdownNodeAction.INSTANCE,
+                new PutShutdownNodeAction.Request(nodeId, type, this.getTestName(), allocationDelay, null, null)
+            )
+        );
+    }
+
+    private void deleteNodeShutdown(String nodeId) {
+        assertAcked(client().execute(DeleteShutdownNodeAction.INSTANCE, new DeleteShutdownNodeAction.Request(nodeId)));
+    }
+
+    private String getNodeId(String nodeName) {
+        NodesInfoResponse nodes = clusterAdmin().prepareNodesInfo().clear().get();
+        return nodes.getNodes()
+            .stream()
+            .map(NodeInfo::getNode)
+            .filter(node -> node.getName().equals(nodeName))
+            .map(DiscoveryNode::getId)
+            .findFirst()
+            .orElseThrow();
+    }
+
+    private void assertNoShuttingDownNodes(String nodeId) throws ExecutionException, InterruptedException {
+        var response = client().execute(GetShutdownStatusAction.INSTANCE, new GetShutdownStatusAction.Request(nodeId)).get();
+        assertThat(response.getShutdownStatuses(), empty());
+    }
+
+    public void testShutdownReadinessService() throws Exception {
+
+        final String nodeName = internalCluster().startNode();
+        final String nodeId = getNodeId(nodeName);
+
+        final var readinessService = internalCluster().getInstance(ReadinessService.class, nodeName);
+
+        // Once we have the right port, check to see if it's ready, has to be for a properly started cluster
+        tcpReadinessProbeTrue(readinessService);
+
+        // Mark the node for shutdown and check that it's not ready
+        putNodeShutdown(nodeId, SingleNodeShutdownMetadata.Type.RESTART, TimeValue.timeValueMillis(1));
```

Comment:
This is the exact porting, but in order to make this more reliable, should we have a greater delay here, so we are sure we can cancel the shutdown? Cancel happens "immediately", but maybe better stay larger here.